example/cli/odp_cli.c

```diff
@@ -0,0 +1,145 @@
+/* Copyright (c) 2021, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/*
+ * ODP CLI Helper Example
+ *
+ * This example shows how to start and stop ODP CLI using the CLI helper
+ * API functions. This example application can also be used to try out
+ * the CLI by connecting to a running application with a telnet client.
+ */
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+#include <odp/helper/cli.h>
+
+#include <stdio.h>
+#include <stdint.h>
+
+typedef struct {
+	int num;
+	char *addr;
+	uint16_t port;
+} options_t;
+
+static void usage(const char *prog)
+{
+	printf("\n"
+	       "Usage: %s [options]\n"
+	       "\n"
+	       "OPTIONS:\n"
+	       "  -n, --num <sec>         Keep CLI open for <sec> seconds. (default 0)\n"
+	       "  -a, --address <addr>    Bind listening socket to IP address <addr>.\n"
+	       "  -p, --port <port>       Bind listening socket to port <port>.\n"
+	       "\n"
+	       "ODP helper defaults are used for address and port, if the options are\n"
+	       "not given.\n"
+	       "\n",
+	       prog);
+}
+
+static void parse_args(int argc, char *argv[], options_t *opt)
+{
+	static const struct option longopts[] = {
+		{ "num", required_argument, NULL, 'n' },
+		{ "address", required_argument, NULL, 'a' },
+		{ "port", required_argument, NULL, 'p' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+n:a:p:h";
+
+	while (1) {
+		int c = getopt_long(argc, argv, shortopts, longopts, NULL);
+
+		if (c == -1)
+			break; /* No more options */
+
+		switch (c) {
+		case 'n':
+			opt->num = atoi(optarg);
+			break;
+		case 'a':
+			opt->addr = optarg;
+			break;
+		case 'p':
+			opt->port = atoi(optarg);
+			break;
+		default:
+			usage(argv[0]);
+			exit(EXIT_SUCCESS);
+			break;
+		}
+	}
+
+	optind = 1; /* reset 'extern optind' from the getopt lib */
+}
+
+int main(int argc, char *argv[])
+{
+	odp_instance_t inst;
+	options_t opt;
+	int i;
+
```

Comment:
Since `odph_cli_start()` creates a new worker, it could be beneficial to support also process mode.

E.g. copy-paste from odp_l2fwd:
```
	odph_helper_options_t helper_options;

	/* Let helper collect its own arguments (e.g. --odph_proc) */
	argc = odph_parse_options(argc, argv);
	if (odph_options(&helper_options)) {
		ODPH_ERR("Error: reading ODP helper options failed.\n");
		exit(EXIT_FAILURE);
	}

	odp_init_param_init(&init);

	init.mem_model = helper_options.mem_model;

	if (odp_init_global(&instance, &init, NULL)) {
...
```